File: kotlin/Checkout/src/CheckoutLine.kt
/**
 * Create a checkout line.
 *
 * @param T concrete type for this generic class.
 */
class CheckoutLine<T>(
    val queue: LinkedQueue<T> = LinkedQueue(),
    private var queueWaitTime: Int = 0
) {
    /**
     * Current wait-time required by the queue (in seconds).
     * This is the total time to serve all the customer within this queue.
     *
     * @return Int time to serve all the customer.
     */
    val waitTime get() = queueWaitTime

    /**
     * Increase the current wait-time of this queue by
     * the time required by the newly added customer.
     *
     * @param timeToServe the time required to serve the customer.
     */
    fun addWaitTime(timeToServe: Int) {
        queueWaitTime += timeToServe
    }

    /**
     * Dump queue data as formated string.
     *
     * @return string representation of the queue.
     */
    override fun toString() = queue.toString()
}

File: flutter/checkout_queue/lib/checkout_line.dart
import 'package:checkout_queue/linked_queue.dart';

/// Create a checkout line of items of type `T`.
class CheckoutLine<T> {
  CheckoutLine();

  final LinkedQueue<T> queue = LinkedQueue();

  int _queueWaitTime = 0;

  // Get current wait-time required by the queue (in seconds).
  // This is the total time to serve all the customer of this queue.
  int get waitTime => _queueWaitTime;

  // Increase the current wait-time of this queue by
  // the time required by the newly added customer.
  void addWaitTime(int timeToServe) => _queueWaitTime += timeToServe;

  // Dump queue data as formated string.
  @override
  String toString() => queue.toString();
}
